nmt.go

```diff
@@ -142,71 +149,119 @@ func (n *NamespacedMerkleTree) ProveRange(start, end int) (Proof, error) {
 	found, proofStart, proofEnd := n.foundInRange(nID)
+
+	// case 2)
 	if !found {
 		// To generate a proof for an absence we calculate the
 		// position of the leaf that is in the place of where
 		// the namespace would be in:
-		proofStart = n.calculateAbsenceIndex(nID)
+		proofStart = n.calculateAbsenceIndex(nID) // TODO [Me] this could simply return a range, to avoid the line below
 		proofEnd = proofStart + 1
 	}
-	// At this point we either found the namespace in the tree or calculated
+
+	// case 3)
+	// At this point we either found leaves with the namespace nID in the tree or calculated
 	// the range it would be in (to generate a proof of absence and to return
 	// the corresponding leaf hashes).
-	n.computeLeafHashesIfNecessary()
+	n.computeLeafHashesIfNecessary()                 // TODO [Me] Why it is needed? cannot we make sure the leaves hashes are calculated as soon as a data is pushed to the tree?
 	proof := n.buildRangeProof(proofStart, proofEnd)
 
 	if found {
 		return NewInclusionProof(proofStart, proofEnd, proof, isMaxNsIgnored), nil
 	}
+	// TODO [Me] the underlying struct type for both inclusion and absence proofs are the same, i.e., `Proof`
+	// TODO [Me] the only distinction is in the presence/absence of the `leafHash`,
+	// TODO [Me] the interpretation of the `Proof` then may be a bit non-trivial
+	// TODO [Me] however, we may want to actually add an extra field to the `Proof`
+	// TODO [Me] that indicates whether the proof is for inclusion or for absence
+
 	return NewAbsenceProof(proofStart, proofEnd, proof, n.leafHashes[proofStart], isMaxNsIgnored), nil
 }
 
+// buildRangeProof returns the nodes (as byte slices) in the range proof of the supplied range i.e.,
+// [proofStart, proofEnd) where proofEnd is non-inclusive.
+// The nodes are ordered according to in order traversal of the namespaced tree. (TODO [Me] I need to double-check the order)
 func (n *NamespacedMerkleTree) buildRangeProof(proofStart, proofEnd int) [][]byte {
-	proof := [][]byte{}
+	// TODO a more secure way would be to make it slice of fixed size arrays i.e., the hash output size
+	proof := [][]byte{} // it is the list of nodes hashes (as byte slices) with no index
 	var recurse func(start, end int, includeNode bool) []byte
+
+	// start, end are indices of leaves in the tree hence they should be within the size of the tree i.e.,
+	// less than or equal to the len(n.leaves)
+	// this recursive algorithm is inspired by RFC6962 https://www.rfc-editor.org/rfc/rfc6962#section-2.1
+	// includeNode indicates whether the hash of the current subtree (covering the supplied range i.e., [start, end)) or
+	// one of its constituent subtrees should be part of the proof
 	recurse = func(start, end int, includeNode bool) []byte {
-		if start >= len(n.leafHashes) {
+		if start >= len(n.leafHashes) { // TODO [Me] why against leafHashes? and not leaves?
+			// TODO [Me] Why there is no check for the end index?
```

Comment:
I _think_ it's not necessary by construction. Maybe the same actually holds for start as well though. Would need to verify that start+k never can exceed len(leaves):
https://github.com/celestiaorg/nmt/blob/29cca3c0d30b05545bf37c09fa19dc070289610a/nmt.go#L211

Reply:
Thanks for clarification, right, `end` can actually exceed the `len(n.leafHashes)` because we call `recurse` on the full tree size i.e., https://github.com/celestiaorg/nmt/blob/ac87f1cd1331b2a6b72b14ab427cffeddccaa154/nmt.go#L288